jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -317,180 +318,129 @@ static ecma_value_t
-
-  JERRY_ASSERT (!ECMA_IS_VALUE_ERROR (new_array_value));
-
-  ecma_object_t *new_array_obj_p = ecma_get_object_from_value (new_array_value);
+  ecma_object_t *new_regexp_obj = ecma_get_object_from_value (new_regexp);
 
-  /* 8.c. */
-  ecma_number_t previous_last_index = 0;
-  /* 8.d. */
-  uint32_t n = 0;
-  /* 8.e. */
-  bool last_match = true;
+  ecma_value_t func_value = ecma_op_object_get_by_symbol_id (new_regexp_obj, LIT_MAGIC_STRING_MATCH);
 
-  ret_value = ECMA_VALUE_ERROR;
+  if (ECMA_IS_VALUE_ERROR (func_value))
+  {
+    ecma_deref_ecma_string (str_p);
+    ecma_deref_object (new_regexp_obj);
+    return func_value;
+  }
 
-  /* 8.f. */
-  while (last_match)
+  if (!ecma_op_is_callable (func_value))
   {
-    /* 8.f.i. */
-    ecma_value_t exec_value = ecma_regexp_exec_helper (regexp_value, this_to_string_value, false);
+    ecma_deref_ecma_string (str_p);
+    ecma_deref_object (new_regexp_obj);
+    ecma_free_value (func_value);
+    return ecma_raise_type_error (ECMA_ERR_MSG ("@@match is not callable."));
+  }
 
-    if (ECMA_IS_VALUE_ERROR (exec_value))
-    {
-      break;
-    }
+  ecma_object_t *func_obj = ecma_get_object_from_value (func_value);
 
-    if (ecma_is_value_null (exec_value))
-    {
-      /* 8.f.ii. */
-      break;
-    }
+  ecma_value_t str_value = ecma_make_string_value (str_p);
 
-    /* 8.f.iii. */
-    ecma_value_t this_index_value = ecma_op_object_get_by_magic_id (regexp_obj_p, LIT_MAGIC_STRING_LASTINDEX_UL);
+  result = ecma_op_function_call (func_obj, new_regexp, &str_value, 1);
 
-    if (ECMA_IS_VALUE_ERROR (this_index_value))
-    {
-      goto cleanup;
-    }
+  ecma_deref_ecma_string (str_p);
+  ecma_deref_object (new_regexp_obj);
+  ecma_deref_object (func_obj);
 
-    ecma_value_t this_index_number = ecma_op_to_number (this_index_value);
+  return result;
+#else /* !ENABLED (JERRY_ES2015) */
 
-    if (ECMA_IS_VALUE_ERROR (this_index_value))
-    {
-      ecma_free_value (this_index_value);
-      goto cleanup;
-    }
+  if (ecma_object_is_regexp_object (regexp_arg))
+  {
+    return ecma_regexp_match_helper (regexp_arg, this_to_string_value);
+  }
 
-    ecma_number_t this_index = ecma_get_number_from_value (this_index_number);
+  ecma_string_t *pattern_p;
 
-    /* 8.f.iii.2. */
-    if (this_index == previous_last_index)
-    {
-      /* 8.f.iii.2.a. */
-      ecma_value_t index_put_value = ecma_op_object_put (regexp_obj_p,
-                                                         ecma_get_magic_string (LIT_MAGIC_STRING_LASTINDEX_UL),
-                                                         ecma_make_number_value (this_index + 1),
-                                                         true);
-      if (ECMA_IS_VALUE_ERROR (index_put_value))
-      {
-        ecma_free_value (this_index_value);
-        ecma_free_number (this_index_number);
-        goto cleanup;
-      }
+  if (ecma_is_value_undefined (regexp_arg))
```

Comment:
You could use `ecma_regexp_read_pattern_str_helper` which handles the toString conversion and the undefined argument.